python/packages/ai/teams/ai/augmentations/tools_augmentation.py

```diff
@@ -0,0 +1,198 @@
+
+    async def validate_response(
+        self,
+        context: TurnContext,
+        memory: MemoryBase,
+        tokenizer: Tokenizer,
+        response: PromptResponse[Union[str, List[chat.ChatCompletionMessageToolCall]]],
+        remaining_attempts: int,
+    ) -> Validation:
+        """
+        Validates a response to a prompt.
+
+        Args:
+            context (TurnContext): Context for the current turn of conversation.
+            memory (MemoryBase): Interface for accessing state variables.
+            tokenizer (Tokenizer): Tokenizer to use for encoding/decoding text.
+            response (PromptResponse[Union[str, List[chat.ChatCompletionMessageToolCall]]]):
+                Response to validate.
+            remaining_attempts (int): Nubmer of remaining attempts to validate the response.
+
+        Returns:
+            Validation: A 'Validation' object.
+        """
+        valid_action_tool_calls: List[chat.ChatCompletionMessageToolCall] = []
+
+        if (
+            response.message
+            and response.message.action_tool_calls
+            and len(response.message.action_tool_calls) > 0
+            and memory.get(SUBMIT_TOOL_OUTPUTS_VARIABLE) is True
+        ):
+            tool_calls = response.message.action_tool_calls
+            tools = self._actions
+
+            if tools and len(tool_calls) > 0:
+                tool_choice = memory.get("temp.tool_choice") or "auto"
+
+                # Calling a single tool
+                if isinstance(tool_choice, dict):
+                    function_name = tool_choice["function"]["name"]
+                    curr_tool_call = tool_calls[0]
+                    curr_tool = None
+
+                    for tool in tools:
+                        if tool.name == function_name:
+                            curr_tool = tool
+                            break
+
+                    if curr_tool:
+                        # Validate function arguments
+                        required_args: List[str] = (
+                            curr_tool.parameters["required"]
+                            if curr_tool.parameters and "required" in curr_tool.parameters
+                            else None
+                        )
+
+                        curr_args = json.loads(curr_tool_call.function.arguments)
+
+                        # Contains all required args
+                        if required_args:
+                            if all(args in curr_args for args in required_args):
+                                valid_action_tool_calls.append(curr_tool_call)
+                        else:
+                            valid_action_tool_calls.append(curr_tool_call)
+                    else:
+                        return Validation(valid=False, feedback="The invoked tool does not exist.")
+
+                # Calling multiple tools
+                else:
+                    for tool_call in tool_calls:
+                        function_name = tool_call.function.name
```

Comment:
instead of duplicating all this logic between the if and the else, you can use the if and the else to initialize an array of tool calls and put this logic into a single look which would result in probably half the lines of code.